python/ml4ir/base/features/feature_fns/categorical.py

```diff
@@ -345,26 +465,55 @@ class VocabLookup(layers.Layer):
     Ref: https://github.com/tensorflow/tensorflow/issues/38305
     """
 
-    def __init__(self, vocabulary_keys, vocabulary_ids, num_oov_buckets, feature_name):
+    def __init__(
+        self,
+        vocabulary_keys,
+        vocabulary_ids,
+        num_oov_buckets: int = None,
+        default_value: int = None,
+        feature_name="categorical_variable",
+    ):
         super(VocabLookup, self).__init__(trainable=False, dtype=tf.int64)
         self.vocabulary_keys = vocabulary_keys
         self.vocabulary_ids = vocabulary_ids
         self.vocabulary_size = len(set(vocabulary_ids))
         self.num_oov_buckets = num_oov_buckets
+        self.default_value = default_value
         self.feature_name = feature_name
 
     def build(self, input_shape):
+        """
+        Defines a Lookup Table  using a KeyValueTensorInitializer to map the keys to the IDs.
+        Allows definition of two types of lookup tables based on whether the user specifies num_oov_buckets or the default_value
+        """
         table_init = lookup.KeyValueTensorInitializer(
             keys=self.vocabulary_keys,
             values=self.vocabulary_ids,
             key_dtype=tf.string,
             value_dtype=tf.int64,
         )
-        self.lookup_table = lookup.StaticVocabularyTable(
-            initializer=table_init,
-            num_oov_buckets=self.num_oov_buckets,
-            name="{}_lookup_table".format(self.feature_name),
-        )
+
+        """
+        NOTE:
+        If num_oov_buckets are specified, use a StaticVocabularyTable
+        otherwise, use a StaticHashTable with the specified default_value
+
+        For most cases, StaticVocabularyTable is sufficient. But when we want to use a custom default_value(like in the case of the dropout function), we need to use StaticHashTable.
+        """
+        if self.num_oov_buckets is not None:
+            self.lookup_table = lookup.StaticVocabularyTable(
+                initializer=table_init,
+                num_oov_buckets=self.num_oov_buckets,
+                name="{}_lookup_table".format(self.feature_name),
+            )
+        elif self.default_value is not None:
+            self.lookup_table = lookup.StaticHashTable(
+                initializer=table_init,
+                default_value=self.default_value,
+                name="{}_lookup_table".format(self.feature_name),
+            )
+        else:
+            raise KeyError("You must specify either num_oov_buckets or default_value")
```

Comment:
@lastmansleeping Technically this error looks impossible to happen to me, since one of them is defined as none in lines 433 and following. Right? But I suppose good to keep this, as a safeguard in case someone changes that code without realizing the waterfall error it would cause.